### PR TITLE
fix: Arrow extension-type metadata was not set when calling the REST API or when there are no rows

### LINKED ISSUE #894: feat: ensure Arrow type metadata is consistent between REST and BQ Storage APIs
The BQ Storage API annotates certain types with additional information so that they can be disambiguated. For example, to tell GEOGRAPHY from STRING columns.

We don't currently annotate the types we create when data is initially downloaded from the REST API.

If we make these consistent, the user could register Arrow extension types, which can hook into the default pyarrow -> pandas conversions. See discussion on: https://github.com/googleapis/python-bigquery/pull/848#issuecomment-904025863

### PR DESCRIPTION
**Pull request by jimfulton**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #894  🦕

### CODE REVIEW

File: google/cloud/bigquery/_pandas_helpers.py [hunk 1]
```diff
@@ -227,7 +234,12 @@ def bq_to_arrow_field(bq_field, array_type=None):
         if array_type is not None:
             arrow_type = array_type  # For GEOGRAPHY, at least initially
         is_nullable = bq_field.mode.upper() == "NULLABLE"
-        return pyarrow.field(bq_field.name, arrow_type, nullable=is_nullable)
+        metadata = BQ_FIELD_TYPE_TO_ARROW_FIELD_METADATA.get(
+            bq_field.field_type.upper() if bq_field.field_type else ""
```
Comment (plamut) [inline google/cloud/bigquery/_pandas_helpers.py:238]: Looking at [pyarrow.field](https://arrow.apache.org/docs/python/generated/pyarrow.field.html#pyarrow.field) docs, `metadata` should be `dict` or `None`, but we sometimes use an empty string. Is there a reason for that?

I suppose it still works, but if nothing else a type checker might start complaining at some point when it realizes that.

File: google/cloud/bigquery/table.py [hunk 2]
```diff
@@ -1810,7 +1810,7 @@ def to_arrow(
             if owns_bqstorage_client:
                 bqstorage_client._transport.grpc_channel.close()
 
-        if record_batches:
+        if record_batches and bqstorage_client is not None:
```
Comment (plamut) [inline google/cloud/bigquery/table.py:1813]: Curious what bug did this extra check for the BQ Storage client fix?

File: tests/system/test_arrow.py [hunk 3]
```diff
@@ -110,3 +110,56 @@ def test_list_rows_nullable_scalars_dtypes(
...
+    dataset_client.query(f"create table {test_table_name} ({columns})").result()
+    if do_insert:
+        names = list(types)
+        values = ", ".join(types[name][1] for name in names)
+        names = ", ".join(names)
+        dataset_client.query(
+            f"insert into {test_table_name} ({names}) values ({values})"
+        ).result()
+    at = dataset_client.query(f"select * from {test_table_name}").result().to_arrow()
+    smd = {at.field(i).name: at.field(i).metadata for i in range(at.num_columns)}
```
Comment (plamut) [inline tests/system/test_arrow.py:152]: (nit)
What does `smd` stand for, Stored metadata? And `rmd` is retrieved metadata? Or is it Storage- and REST- metadata, respectively?

Had some trouble deciphering that, maybe find a better name? :)

**Review by plamut — COMMENTED** [review]
A few nits/questions, but generally looks good to me.

File: google/cloud/bigquery/_pandas_helpers.py (hunk 1 above)
Comment (jimfulton) [reply google/cloud/bigquery/_pandas_helpers.py:238]: The empty string is just the argument to `get`. It's not metadata.  We only pass a `dict` or `None`.

File: google/cloud/bigquery/table.py (hunk 2 above)
Comment (jimfulton) [reply google/cloud/bigquery/table.py:1813]: When `bqstorage_client` is `None`, we're using the REST API, which doesn't add arrow extension metadata.

I improved the comment.

File: tests/system/test_arrow.py (hunk 3 above)
Comment (jimfulton) [reply tests/system/test_arrow.py]: > (nit)
> What does `smd` stand for, Stored metadata? And `rmd` is retrieved metadata? Or is it Storage- and REST- metadata, respectively?

The later.

> Had some trouble deciphering that, maybe find a better name? :)

Good idea. Done.

File: google/cloud/bigquery/_pandas_helpers.py (hunk 1 above)
Comment (plamut) [reply google/cloud/bigquery/_pandas_helpers.py:238]: Meh, of course, and an empty string won't mach any keys, resulting in `None`. And I though I saw a comma somewhere making `""` a second argument. Nevermind. :)

**Review by plamut — APPROVED** [review]
LGTM!

**Review by tswast — COMMENTED** [review]
Thanks

**Pull request merged**